### PR TITLE
changed TE 'Allocate/Free' API to take argument 'Buf' instead of arguments 'Var', 'dtype', 'dims'.

### PR DESCRIPTION
**Pull request by huiguoo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50167 changed TE 'Allocate/Free' API to take argument 'Buf' instead of arguments 'Var', 'dtype', 'dims'.**

Differential Revision: [D25814342](https://our.internmc.facebook.com/intern/diff/D25814342)

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/stmt.h [hunk 1]
```diff
@@ -319,35 +322,24 @@ class TORCH_API Allocate : public StmtNode<Allocate> {
...
-      Dtype dtype,
-      const std::vector<const Expr*>& dims)
-      : buffer_var_(buffer_var), dtype_(dtype), dims_(dims) {}
-
+  Allocate(Buf* buf) : buf_(buf) {}
  private:
-  const Var* buffer_var_;
-  Dtype dtype_;
-  std::vector<const Expr*> dims_;
+  Buf* buf_;
```
Comment (huiguoo) [inline torch/csrc/jit/tensorexpr/stmt.h:342]: @ZolotukhinM I am not sure whether it is correct to replace the three private variables with `buf_`. It makes more sense to me but to make it work I have to change the return value of `class Allocate::dims()` to `const vector` from `const vector&`. Guess it will probably introduce some overhead on the performance. 

Wondering why `class Buf::dims()` returns `vector` instead of `const vector&` in the first place.

**Review by ZolotukhinM — COMMENTED** [review]
Good work! I left some comments inline, please take a look. Also, do you plan to convert `Free` in a separate PR? It might be worth changing them both for consistency.

**Comment by huiguoo** [comment]
> Good work! I left some comments inline, please take a look. Also, do you plan to convert `Free` in a separate PR? It might be worth changing them both for consistency.

Thank you! This PR is in early stages and I will convert `Free` in the same PR when the conversion for `Allocate` is complete. The title is a bit misleading - will change it :)

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 2]
```diff
@@ -1988,7 +1988,7 @@ Stmt* TermExpander::mutate(const Allocate* v) {
     return (Stmt*)v;
   }
 
-  return new Allocate(buffer_var_new, v->dtype(), dims_new);
+  return new Allocate(new Buf(buffer_var_new, dims_new, v->dtype()));
```
[the remark this answers is not in the document]
Comment (huiguoo) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp]: It turns out that the flattening code can be extracted to a helper function and shared by the two :) At the same time, I added the `TermExpander::mutate` function for `Buf*` so `TermExpander::mutate(Allocate*)` is more neat.

[the remark this answers is not in the document]
Comment (huiguoo) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp]: Update: `TermExpander::mutate(const Buf*)` is removed - it caused lots of failures in cpp tests.

File: test/cpp/tensorexpr/test_simplify.cpp [hunk 3]
```diff
@@ -3554,11 +3556,13 @@ TEST(Simplify, SimplifyEliminateZeroLengthAlloc) {
     // Finds right Alloc/Free.
     VarHandle x("x", kInt);
     VarHandle y("y", kInt);
+    BufHandle b1(x, {0}, kInt);
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_simplify.cpp:3559]: I think `Var` used as base handles in `Buf`s need to have `kHandle` dtype (not `kInt)`. But maybe we could remove them here entirely? I.e. replace
```
    VarHandle x("x", kInt);
    BufHandle b1(x, {0}, kInt);
```
with
```
    BufHandle b1("x", {0}, kInt);
```

File: test/cpp/tensorexpr/test_simplify.cpp [hunk 4]
```diff
@@ -3577,11 +3581,13 @@ TEST(Simplify, SimplifyEliminateZeroLengthAlloc) {
     VarHandle x("x", kInt);
     VarHandle y("y", kInt);
     VarHandle z("z", kInt);
+    BufHandle b1(x, {0}, kInt);
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_simplify.cpp:3584]: Ditto.

File: torch/csrc/jit/tensorexpr/expr.h [hunk 5]
```diff
@@ -165,13 +165,45 @@ class Var : public ExprNode<Var> {
...
+};
+
 class TORCH_API Buf : public ExprNode<Buf> {
  public:
   static ExprHandle make(
       const std::string& name_hint,
       const std::vector<ExprHandle>& dims,
       Dtype dtype);
   static ExprHandle make(const std::vector<ExprHandle>& dims, Dtype dtype);
+  static ExprHandle make(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/expr.h:203]: I'm not sure this is necessary. We could create `Buf` objects directly without first creating `Var`s for them - these vars are usually not used anywhere.

File: torch/csrc/jit/tensorexpr/ir.cpp [hunk 6]
```diff
@@ -144,6 +145,14 @@ Store* Store::make(
       ExprHandle(1).node());
 }
 
+Allocate* Allocate::make(
+    const VarHandle& buffer_var,
+    Dtype dtype,
+    const std::vector<ExprHandle>& dims) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir.cpp:151]: I think we should remove API for creating `Allocate` statements with `Var` - the API taking `Buf` should handle everything we need.

File: torch/csrc/jit/tensorexpr/ir_mutator.cpp [hunk 7]
```diff
@@ -186,7 +186,22 @@ const Expr* IRMutator::mutate(const Load* v) {
 }
 
 const Expr* IRMutator::mutate(const Buf* v) {
-  return v;
+  const Var* var = v->base_handle();
+  const Var* var_new = dynamic_cast<const Var*>(var->accept_mutator(this));
+  bool any_change = var_new != var;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_mutator.cpp:191]: Theoretically, `var_new` might be `nullptr` after `dynamic_cast`. We probably should handle this case - i.e. return `nullptr` (returning `Buf(var_new, dims_new, v->dtype())` would produce a partially invalid IR, which is can lead to hard to debug issues).

File: torch/csrc/jit/tensorexpr/ir_mutator.cpp [hunk 8]
```diff
@@ -406,34 +421,22 @@ Stmt* IRMutator::mutate(const SyncThreads* v) {
...
-
-  if (!any_change) {
+  const Buf* buf = v->buf();
+  const Buf* buf_new = dynamic_cast<const Buf*>(buf->accept_mutator(this));
+  if (buf == buf_new) {
     return (Stmt*)v;
   }
-
-  return new Allocate(buffer_var_new, v->dtype(), dims_new);
+  return new Allocate(buf_new);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_mutator.cpp:429]: Similarly, we need to check (or assert) that `buf_new` is not nullptr.

File: torch/csrc/jit/tensorexpr/ir_mutator.cpp [hunk 9]
```diff
@@ -515,11 +518,11 @@ Stmt* StmtClone::mutate(const AtomicAdd* v) {
 }
 
 Stmt* StmtClone::mutate(const Allocate* v) {
-  return new Allocate(v->buffer_var(), v->dtype(), v->dims());
+  return new Allocate(new Buf(v->buffer_var(), v->dims(), v->dtype()));
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_mutator.cpp:521]: If we had `Allocate` and `Free` statements accessing the same `Buf` and clone both of them, the clones would refer to different `Buf`s, because when we clone `Allocate` we create a new `Buf` object. We should be using `v->buf()` instead here - `Buf` are expressions and expressions are immutable, so we don't need to create copies of them (in contrast to statements, which are mutable, and for which we do need to create copies).

File: torch/csrc/jit/tensorexpr/ir_simplifier.h [hunk 10]
```diff
@@ -610,6 +609,8 @@ class TORCH_API IRSimplifier {
   }
 };
 
+// Flattens the buf and performs the simplifier on the flattened dims.
+const Expr* buf_flattening_helper(const Buf* v);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_simplifier.h:613]: Nit: this function might benefit from a more specific name.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 11]
```diff
@@ -1961,23 +1961,34 @@ const Expr* TermExpander::mutate(const RoundOff* v) {
...
   std::vector<const Expr*> dims_new(dims_old.size());
   for (size_t i = 0; i < dims_old.size(); i++) {
     dims_new[i] = dims_old[i]->accept_mutator(this);
     any_change |= (dims_new[i] == dims_old[i]);
-    flattened = new Mul(flattened, dims_new[i]);
   }
 
   // Safe to do this as there can't be an Allocate inside an Allocate:
-  flattened = IRSimplifier::simplify(flattened);
+  const Buf* buf_new = new Buf(buffer_var_new, dims_new, v->dtype());
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:1990]: Could we replace this with simply mutating the `Buf`?
```
const Buf* buf_new = v->buf()->accept_mutator(this);
```

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 12]
```diff
@@ -1988,24 +1999,23 @@ Stmt* TermExpander::mutate(const Allocate* v) {
...
 
   if (eliminated_allocations_.count(buffer_var_new)) {
     eliminated_allocations_.erase(buffer_var_new);
     return nullptr;
   }
 
-  if (buffer_var_new == buffer_var_old) {
+  if (buffer_var_new == buffer_var) {
     return (Stmt*)v;
   }
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:2017]: Ditto.

File: torch/csrc/jit/tensorexpr/loopnest.cpp [hunk 13]
```diff
@@ -940,13 +940,10 @@ Stmt* LoopNest::insertAllocFree(Stmt* stmt) {
...
-    for (auto& d : buf->dims()) {
-      flat_size = new Mul(flat_size, d);
-    }
-    flat_size = IRSimplifier::simplify(flat_size);
-    Stmt* alloc = new Allocate(buf->base_handle(), buf->dtype(), {flat_size});
-    Stmt* free = new Free(buf->base_handle());
+    const Expr* flat_size = buf_flattening_helper(buf);
+    Stmt* alloc =
+        new Allocate(new Buf(buf->base_handle(), {flat_size}, buf->dtype()));
+    Stmt* free = new Free(buf);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/loopnest.cpp:946]: This `Free` refers to a different `Buf` object.

**Review by ZolotukhinM — COMMENTED** [review]
Sorry for the delayed review, I didn't realize it was waiting for me! It looks very good, I have just a couple of comments. Once they are addressed, this could be landed!

Thank you for cleaning this part of our system!

File: torch/csrc/jit/tensorexpr/ir_simplifier.h (hunk 10 above)
Comment (huiguoo) [reply torch/csrc/jit/tensorexpr/ir_simplifier.h:613]: Any suggestions? :)

File: torch/csrc/jit/tensorexpr/loopnest.cpp [hunk 14]
```diff
@@ -940,13 +940,10 @@ Stmt* LoopNest::insertAllocFree(Stmt* stmt) {
...
     Block* alloc_block = findLowestContainingBlock(uses.at(buf));
```
Comment (huiguoo) [inline torch/csrc/jit/tensorexpr/loopnest.cpp:947]: FYI, when using `buf_new` instead of `buf` for this line, there are lots of failing cpp reduction tests. Didn't investigate on it though.

File: torch/csrc/jit/tensorexpr/ir_simplifier.h (hunk 10 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/ir_simplifier.h:613]: Maybe `buf_flat_size` or something like that reflecting what this function is computing?

File: torch/csrc/jit/tensorexpr/loopnest.cpp (hunk 14 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/loopnest.cpp:947]: We probably need to rewrite all uses of `buf` with `buf_new` before doing that. We created a new buffer, but there might be statements still referring to the old one (it wasn't the case before this change since the buf stayed the same). We probably could do that using `Substitute`, but probably it needs to be expanded to support Buf-Buf mappings:
https://github.com/pytorch/pytorch/blob/16cfe970e0dd3c7eada27ef533ed566c55850020/torch/csrc/jit/tensorexpr/eval.h#L225

File: torch/csrc/jit/tensorexpr/expr.h [hunk 15]
```diff
@@ -165,6 +165,34 @@ class Var : public ExprNode<Var> {
   std::string name_hint_;
 };
 
+// An expression to construct the underlying variable node.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/expr.h:168]: Why was this moved? If there is no real need in that, it's better to avoid unnecessary code movements.

File: torch/csrc/jit/tensorexpr/ir.cpp [hunk 16]
```diff
@@ -1,5 +1,6 @@
 #include <torch/csrc/jit/tensorexpr/ir.h>
 
+#include <torch/csrc/jit/tensorexpr/stmt.h>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir.cpp:3]: This seems unnecessary, could we undo that?

File: torch/csrc/jit/tensorexpr/ir_mutator.cpp [hunk 17]
```diff
@@ -187,7 +187,24 @@ const Expr* IRMutator::mutate(const Load* v) {
 }
 
 const Expr* IRMutator::mutate(const Buf* v) {
-  return v;
+  const Var* var = v->base_handle();
+  const Var* var_new = dynamic_cast<const Var*>(var->accept_mutator(this));
+  if (!var_new)
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_mutator.cpp:192]: Styling nit: please enclose the if body with {}.

File: torch/csrc/jit/tensorexpr/ir_mutator.cpp [hunk 18]
```diff
@@ -425,34 +442,24 @@ Stmt* IRMutator::mutate(const ExternalCall* v) {
...
-  std::vector<const Expr*> dims_new(dims_old.size());
-  for (size_t i = 0; i < dims_old.size(); i++) {
-    dims_new[i] = dims_old[i]->accept_mutator(this);
-    any_change |= (dims_new[i] != dims_old[i]);
-  }
-
-  if (!any_change) {
+  const Buf* buf = v->buf();
+  const Buf* buf_new = dynamic_cast<const Buf*>(buf->accept_mutator(this));
+  assert(buf_new);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_mutator.cpp:447]: Nit: I think we should use `TORCH_INTERNAL_ASSERT` for asserts.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 19]
```diff
@@ -1981,51 +1981,51 @@ const Expr* TermExpander::mutate(const RoundOff* v) {
...
-  // Safe to do this as there can't be an Allocate inside an Allocate:
   flattened = IRSimplifier::simplify(flattened);
 
+  return flattened;
+}
+
+Stmt* TermExpander::mutate(const Allocate* v) {
+  const Buf* buf = v->buf();
+  const Buf* buf_new = dynamic_cast<const Buf*>(v->buf()->accept_mutator(this));
+  assert(buf_new);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:1999]: Ditto.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 20]
```diff
@@ -1981,51 +1981,51 @@ const Expr* TermExpander::mutate(const RoundOff* v) {
...
+  return new Allocate(buf_new);
 }
 
 Stmt* TermExpander::mutate(const Free* v) {
-  const Expr* buffer_var_old = v->buffer_var();
-  const Var* buffer_var_new =
-      dynamic_cast<const Var*>(buffer_var_old->accept_mutator(this));
+  const Buf* buf = v->buf();
+  const Buf* buf_new = dynamic_cast<const Buf*>(v->buf()->accept_mutator(this));
+  assert(buf_new);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:2017]: Ditto.

**Review by ZolotukhinM — APPROVED** [review]
This looks great, thank you for cleaning this up! I left some minor remarks, please address them before landing.

File: torch/csrc/jit/tensorexpr/expr.h (hunk 15 above)
Comment (huiguoo) [reply torch/csrc/jit/tensorexpr/expr.h:168]: Oops, I added one constructor for `Buf `that uses `VarHandle` in prior commits so moved it ahead of `Buf`. The constructor turns out unnecessary and was removed in later commits. Forgot to move `VarHandle` back.
